src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,266 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+#pragma unroll
+      for (index_t c_dim_id = 0; c_dim_id < cl_elems; ++c_dim_id) {
+        sum = cl::sycl::mad(matrix_scratch[mat_index], *vector_scratch++, sum);
+
+        mat_index += local_range + 1;  // Adding one as bank offset
+      }
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+
+    if (nc_dim_index < nc_dim) {
+      lhs_.eval(out_index) = sum;
+    }
+    return sum;
+  }
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          bool is_transposed, int cache_line_size, int work_per_thread>
+template <typename ScratchPointerType>
+SYCL_BLAS_INLINE void
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size,
+     work_per_thread>::extract_input_block(ScratchPointerType matrix_scratch,
+                                           const index_t &local_id,
+                                           const index_t &group_id,
+                                           const index_t &local_range,
+                                           const index_t &lda,
+                                           index_t c_tile_id) {
+  constexpr int cl_elems = cache_line_size / sizeof(value_t);
+
+  const index_t nc_dim =
+      is_transposed ? matrix_a_.get_size_col() : matrix_a_.get_size_row();
+  const index_t c_dim =
+      is_transposed ? matrix_a_.get_size_row() : matrix_a_.get_size_col();
+
+  // Group indices within global grid
+  const index_t nc_group_id = group_id / WGs_per_C_;
+  const index_t c_group_id = group_id - nc_group_id * WGs_per_C_;
+
+  // Tile dimensions
+  const int tile_dim_nc = local_range / cl_elems;
+  constexpr int tile_dim_c = cl_elems;
+
+  // In-tile indices
+  const index_t tile_nc_index = local_id / tile_dim_c;
+  const index_t tile_c_index = local_id % tile_dim_c;
+
+  // Tile-group (local) indices
+  index_t local_nc_index = tile_nc_index;
+  const index_t local_c_index = tile_c_index + c_tile_id * tile_dim_c;
+
+  // Global (grid) indices
+  const index_t grid_nc_index = local_nc_index + nc_group_id * local_range;
+  const index_t grid_c_index = local_c_index + c_group_id * local_range;
+
+  // Matrix index calculated based on global grid indices
+  index_t mat_index = grid_c_index + grid_nc_index * lda;
+
+  // Number of iterations to load all tiles
+  // local_range / tile_dim_nc == cl_elems
+  constexpr int tile_nc_loops = cl_elems;
+
+  // Transposed index in scratch-space with bank offset which adds 1 word of
+  // padding every 'tile_dim_c' elements
+  const index_t scratch_index = (local_range + 1) * tile_c_index;
+
+#pragma unroll
+  for (index_t nc_tile_id = 0; nc_tile_id < tile_nc_loops; ++nc_tile_id) {
+    // Populate the scratch memory with matrix values, ensuring we don't fetch
+    // beyond bounds
+    matrix_scratch[scratch_index + local_nc_index] =
+        grid_nc_index < nc_dim && grid_c_index < c_dim
```

Comment:
This is still wrong. `grid_c_index` and `grid_nc_index` are constants but the actual nc index is changing with `mat_index += tile_dim_nc * lda`.

Reply:
Ah you're right. I will fix it soon.